Add ERC: Agent Tool Registry (#1723)
ethereum/ERCs: ERCS/erc-8257.md

**Onchain commitment matters.** By storing a `keccak256` hash of the manifest onchain, consumers can verify that the manifest they fetched has not been tampered with. Combined with origin-binding, this provides a lightweight trust anchor without requiring onchain access checks or gateway infrastructure.

**Pricing is part of discovery.** An agent choosing between two tools that do the same thing needs to know cost. Declaring pricing in the manifest (and committing it by hash onchain) lets consumers compare tools before invocation. The manifest declares what the tool costs; the endpoint enforces payment. The registry itself never handles funds. The pricing schema is deliberately protocol-agnostic: it identifies the payment protocol by an opaque string so the standard does not depend on any specific payment system.

## Specification

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

Reviewer: ```suggestion
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).
```

> This ERC targets EVM chains. The registry interface uses EVM-native types (`address`, `bytes32`, `uint256`), and tool IDs are scoped to the `(chainId, registryAddress)` tuple of an EVM deployment (see §1). Pricing entries carry their own chain identifiers via CAIP-19 `asset` and CAIP-10 `recipient`, so a tool registered on an EVM chain MAY price itself in assets on any CAIP-10/CAIP-19 namespace; concrete guidance in §3 focuses on `eip155:*` because that is the namespace the reference implementation has been exercised against.

Reviewer: Could you link to the CAIPs?

The `manifestHash` in `ToolConfig` is `keccak256` over the **JSON Canonicalization Scheme (JCS)** form of the manifest, as defined by [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785). JCS normalizes object key order, whitespace, number formatting, and string escaping so that semantically identical manifests produce identical byte sequences. Implementations MUST canonicalize with JCS before hashing, and consumers MUST canonicalize before verifying. Servers MAY serve the manifest in any JSON form; the hash commits to the JCS form regardless of transport representation.

JCS does not normalize Unicode string content, hex-digit case, or byte-order marks, so three extra rules apply to every manifest before JCS is run:

- **Unicode normalization:** all JSON string values MUST be in Unicode Normalization Form C (NFC) per Unicode 16.0 or later. Producers MUST NFC-normalize before JCS serialization; consumers MUST reject a fetched manifest whose strings are not already NFC-normalized (no silent re-normalization, since that would change the bytes that were hashed). This prevents hash divergence between producers that canonicalize with NFC and consumers that do not.

> - **Byte-order mark:** the manifest MUST be served as UTF-8 without a byte-order mark. Consumers MUST reject a fetched response whose bytes begin with `EF BB BF` rather than silently stripping the BOM, because silent stripping would change the bytes fed to `keccak256` and cause a hash mismatch that is difficult to diagnose.

Reviewer: ```suggestion
- **Byte-order mark (BOM):** the manifest MUST be served as UTF-8 without a byte-order mark. Consumers MUST reject a fetched response whose bytes begin with `EF BB BF` rather than silently stripping the BOM, because silent stripping would change the bytes fed to `keccak256` and cause a hash mismatch that is difficult to diagnose.
```

- **Hex-field casing:** every hex string in the manifest (`creatorAddress`, the `0x…` portion of any CAIP-19 `asset` or CAIP-10 `recipient`, `access[].requirements[].kind`, `access[].requirements[].data`, `verifiability.attestation.enclaveHash`, `verifiability.reproducibleBuild.buildHash`, and any future hex-string field added by this ERC) MUST use lowercase hex digits. JCS does not case-fold hex, so two manifests that differ only in hex case produce different `manifestHash` values. Producers MUST emit lowercase hex; consumers MUST reject manifests containing uppercase hex digits in any of the listed fields rather than silently lowercasing them, since silent lowercasing would change the bytes fed to `keccak256` and defeat the hash commitment. The per-field grammar is pinned in each field's row in §2 (`creatorAddress` in [Required Fields](#required-fields), `kind` / `data` in [§4 Access](#4-access), `enclaveHash` / `buildHash` in [§5 Verifiability](#5-verifiability)).

> All three rules are treated as verification failures (see [§7 Handling Verification Failure](#handling-verification-failure)). Concrete hash-divergence vectors for NFC-vs-NFD and with-vs-without-BOM appear in [Test Cases](#test-cases) alongside canonical reference manifests.

Reviewer: NFD is not defined as abbreviation, could you define it?

The `metadataURI` declared onchain MUST exactly equal this URL once both sides have been normalized per the rules below; consumers MUST NOT accept a manifest served from any other location.

Origin comparison follows [RFC 6454](https://www.rfc-editor.org/rfc/rfc6454): same scheme, host, and port. `endpoint` MAY include a path or query string; only its scheme, host, and port participate in the origin check.

#### URL Normalization

Because URLs have multiple equivalent representations, both registrants and consumers MUST reduce HTTPS URLs in this ERC to a canonical form before writing, reading, or comparing them. The rules split into two groups: a **general HTTPS-URL** group that applies to every URL field (notably `endpoint` and `image`), and a **well-known-path** group that applies additionally to `metadataURI`.

**General HTTPS-URL normalization** (rules G1‑G3):

- **G1.** Lowercase the scheme and host.
- **G2.** Omit port 443 (the default for `https`).

> - **G3.** Normalize the host as an A-label (ASCII Compatible Encoding, per [RFC 5891](https://www.rfc-editor.org/rfc/rfc5891)) before lowercasing. Consumers MUST reject a URL whose host is given as a U-label (internationalized) without ACE encoding.

Reviewer: ```suggestion
- **G3.** Normalize the host as an A-label (ASCII Compatible Encoding (ACE), per [RFC 5891](https://www.rfc-editor.org/rfc/rfc5891)) before lowercasing. Consumers MUST reject a URL whose host is given as a U-label (internationalized) without ACE encoding.
```

| `ISubscription` | `bytes4(keccak256("subscription()"))` | `0x44387cc2` |

Predicates that compose existing requirements MAY reuse these IDs verbatim. Predicates that introduce a new requirement type MUST publish a marker interface following the same single-zero-argument-function pattern, document the `data` layout, and verify that the resulting selector does not collide with any ID in this section or any prior published marker.

## Reference Implementation

A Foundry-based reference implementation of `ToolRegistry`, `IToolRegistry`, and `IAccessPredicate` is maintained alongside this ERC. It pins the ERC-165 interface id, enforces the manifest-URI length cap described in Security Considerations, and exercises the predicate integration path (ERC-165 validation, non-canonical bool rejection, zero-code predicate handling, gas-capped `staticcall`) through its test suite. Consumers and implementers can use it as a conformance baseline; any downstream implementation is expected to reproduce the pinned interface id and the behavior validated by the test suite.

## Security Considerations

### Predicate Gas Consumption

> The registry delegates access checks to an external predicate contract via `staticcall`. A malicious or poorly written predicate could consume unbounded gas, causing `hasAccess` calls to revert or become prohibitively expensive. Implementations MUST cap the gas forwarded to predicate calls at no more than 200,000 gas (i.e., `staticcall{gas: <=200_000}`) and MUST treat out-of-gas as "access denied." Without a normative ceiling, onchain consumers (paymasters, gated execution routers, contracts that compose `hasAccess` into their own logic) cannot reason about worst-case gas and the registry becomes unsafe to compose without per-deployment auditing.

Reviewer: Review comment: this hardcodes gas sent and this is generally not a good idea because of possible future gas repricings.

Implementations MAY choose a smaller cap. The 200,000-gas value is the upper bound: implementations exceeding it are non-conformant. The cap is chosen so that realistic predicate logic completes comfortably below it:

> - An ERC-20 or [ERC-721](./eip-721.md) `balanceOf` check plus a comparison runs in roughly 5,000-10,000 gas.

Reviewer: review comment: These specific gas mentions should also be removed IMO because this depends on the current fork and also on the EVM bytecode produced by the compiler used (including compiler settings like optimizations)

- Merkle proof verification over a depth-20 tree (supporting ~1,000,000 leaves) runs in roughly 40,000-60,000 gas using OpenZeppelin's `MerkleProof` library.
- An ECDSA signature verification (`ecrecover` plus constant work) runs in roughly 5,000-10,000 gas.

> A 200,000 gas cap leaves headroom of at least 2-3× over the most expensive common case (Merkle proof verification), which keeps the happy path well below the cap while still bounding worst-case gas so that a malfunctioning predicate cannot turn `hasAccess` into an expensive storage-depleting call. Implementations that only expect trivial gates MAY choose a smaller cap. Implementations that expect more elaborate predicates (e.g., zero-knowledge proof verification) MUST instead deploy a derivative registry that explicitly opts out of `IToolRegistry` conformance (a new interface ID with a documented higher cap), since a higher cap weakens the worst-case-gas guarantee that downstream callers rely on. In all cases the outcome of a capped `staticcall` is the same: on OOG the call reverts, and the registry treats the revert as "access denied" (or, via `tryHasAccess`, as `(ok=false, granted=false)`).

Reviewer: OOG -> out of gas

Predicate authors and callers also need to account for [EIP-150](./eip-150.md)'s 63/64 rule: a staticcall forwards at most `63/64` of the caller's remaining gas, so the gas the predicate actually receives is `min(cap, floor(63/64 * gasLeft))`. A caller that invokes `hasAccess` with too little remaining gas can therefore starve a predicate that would otherwise succeed, causing a spurious "access denied" result. Predicate authors SHOULD budget the common path well below the cap (a practical rule of thumb is to keep happy-path execution under half of the cap) so minor variations in caller gas do not flip the outcome. Contracts that invoke `hasAccess` onchain SHOULD ensure they enter the call with at least `cap * 64 / 63` gas available.

### Predicate Upgradeability

> If the `accessPredicate` is a proxy contract (e.g., an [ERC-1967](./eip-1967.md) transparent proxy or UUPS proxy), the predicate owner can silently change the access logic without the tool creator calling `setAccessPredicate`. This means tool consumers cannot rely solely on the `AccessPredicateUpdated` event to detect changes in access semantics. Consumers SHOULD check whether a predicate address contains proxy patterns (e.g., ERC-1967 storage slots) and SHOULD treat upgradeable predicates as higher risk than immutable ones.

Reviewer: UUPS proxy abbreviation not defined

- Pin the expected bytecode hash (or the deployment transaction) for any registry treated as canonical.
- Publish and cross-check the canonical registry address per chain via a well-known directory (e.g., a pinned value in each discovery layer's configuration).

Discovery layers (indexers, agent frameworks, wallets) SHOULD publish the registry bytecode hash alongside the registry address so downstream surfaces can verify the deployment has not been swapped. A registry deployed directly (no proxy) with its source verified on the canonical block explorer is the RECOMMENDED configuration.

### Predicate Reverting

Implementations MUST treat a reverting predicate call as "access denied" from `hasAccess` rather than bubbling the revert to the caller; this prevents a malfunctioning predicate from breaking the registry's view functions. Consumers that need to distinguish a clean denial from a malfunction SHOULD use `tryHasAccess` instead, which reports the predicate outcome as `(ok, granted)`: a malfunction surfaces as `(false, false)` while a clean denial surfaces as `(true, false)`.

### Predicate Reentrancy

> Although `hasAccess` is a `view` function, the predicate it delegates to is an external contract. Implementations MUST call the predicate via `staticcall`, which the EVM forbids from mutating state. State-mutating entrypoints on the registry make no external calls to the predicate at all, so the registry exposes no reentrancy surface for state writes.

Reviewer: Nit: I would drop `view` here and instead directly refer the staticcall. View is a solidity concept, staticcall the EVM concept.

> Read-only reentrancy is a separate hazard that the staticcall guarantee does not eliminate. A predicate is permitted to read state from any third contract; if a consumer calls `hasAccess` while another protocol's view of its own state is mid-transition (e.g., during a multi-call that updates balances between calls, or via a callback into the consumer mid-execution), the predicate may observe stale or inconsistent state and return a `granted` answer that is no longer true once the outer transaction settles. This is the same class of bug that affected several DeFi protocols whose oracles read view state from contracts mid-mutation. Consumers that gate state changes on `hasAccess` MUST NOT call it between writes whose intermediate state another protocol's view function depends on, and SHOULD apply the same reentrancy guards to `hasAccess` invocations that they apply to any other external view call whose inputs include third-party state.

Reviewer: This is an interesting pattern, is there some way to already do this? You can share memory between call frames using transient storage and you could thus save the hasAccess on the first time that this contract is being called (top level, but not necessarily call stack top level, which could thus still mean that once tx finishes the hasAccess is now false)

Predicates and downstream enforcers (tool endpoints, wallets, agent frameworks, contracts that gate behavior on the result) MUST NOT treat a `true` return value as proof that the current requester is `account`. Enforcers MUST independently bind `account` to the real principal before acting on a positive answer, for example by:

- checking `account == msg.sender` when the predicate is consulted from within a transaction initiated by `account`,
- requiring the caller to present a signature over a challenge in `data`, verified by the predicate or the endpoint,
- issuing a short-lived session token after an out-of-band authentication step.

A predicate that gates purely on `account` (e.g., "is this address a holder of NFT X?") is safe to consult but unsafe to act on without such binding. Ignoring this distinction is the most common way that correct-looking access gates become unsound.

#### Concrete AccessProof Pattern

The following challenge-response pattern binds the `account` parameter to a real principal and prevents replay across tools and time windows. Implementations that need authenticated access SHOULD use this pattern or an equivalent that provides the same properties.

> 1. **Challenge issuance.** The agent or consumer generates a challenge: `challenge = keccak256(abi.encodePacked(toolId, account, block.chainid, deadline))`, where `deadline` is a Unix timestamp after which the proof expires.

Reviewer: This might be a security concern. You will have to check if there are no ways someone creates a malicious challenge where the signer signs something which is also valid in another context. For instance: signing a transaction, or something what comes close (this is a list of 4 items which could maybe be tricked here): https://eips.ethereum.org/EIPS/eip-7702 the authorizatoin part where `keccak(MAGIC || rlp([chain_id, address, nonce]))` is signed.
The usual way to mitigate this is to prefix it with a non-used prefix, but these prefix above is used on the core protocol and am not sure if there is an ERC way to safe-prefix these signers to ensure nothing is signed which could be used in another context. ERC-prefixed-signing. 
You could also use a prefixed string for instance used in `eth_sign` to prevent collissions: https://ethereum.org/developers/docs/apis/json-rpc/#eth-sign

(please check with a security expert)

### Zero-Code Access Predicates

Implementations of `IToolRegistry.registerTool` and `IToolRegistry.setAccessPredicate` MUST treat addresses with no deployed code (externally-owned accounts, or CREATE2 addresses that have not yet been deployed) as accepted but unverifiable, per step 2 of [Predicate Validation at Registration](#predicate-validation-at-registration). ERC-165 cannot be queried against empty code, so such a predicate cannot be checked at registration time.

At invocation time, a staticcall to a zero-code address returns empty data, which the registry MUST treat as non-compliant (see [§1 `hasAccess`](#itoolregistry-interface)) and therefore as "access denied." A tool registered with a zero-code predicate is consequently inaccessible until a contract is deployed at that address.

Creators who rely on counterfactual deployment (registering a predicate before deploying it) SHOULD:

> 1. Verify that the CREATE2 salt and init-code commit to the intended predicate bytecode before registration.

Reviewer: The predicate bytecode is here the runtime bytecode (deployed bytecode) I assume? This could be crafted malicious, for instance in the initcode you do SSTORE(0, X). In the runtime bytecode the code looks like:

```
if SLOAD(0) == 0:
- do normal stuff
else
- do malicious predicate
``` 

If we commit to the runtime bytecode above we could thus in practice have two code paths which depend on whatever storage was stored in the initcode. So two different initcodes create the same runtime code, but in practice it will only be able to run one of the two codes

Implementations of `registerTool` and `setAccessPredicate` MUST validate the candidate `accessPredicate` per the following best-effort ERC-165 ladder. The reference implementation pins these rules in `_validatePredicate`; downstream implementations MUST reproduce the same accept/reject behavior (the precise revert path may use an equivalent error per the `InvalidAccessPredicate` docstring).

1. If the predicate is `address(0)`, the call is open-access. Skip validation and accept.
2. If the predicate has no deployed code (`extcodesize == 0`), accept as best-effort. ERC-165 cannot be queried against empty code, so the registration is recorded but the tool is inaccessible until a contract is deployed at the address (see [Zero-Code Access Predicates](#zero-code-access-predicates)).

> 3. Probe `IERC165(predicate).supportsInterface(type(IERC165).interfaceId)` with a strict gas cap. The reference implementation uses 30,000 gas, which matches the ERC-165 standard's documented ceiling (the standard itself requires `supportsInterface` to use *less than* 30,000 gas, so a conformant probe completes within the cap and a probe that exhausts it is non-conformant by construction). Implementations MAY choose any cap between 1,000 and 30,000 gas; a cap below 1,000 is too tight for a realistic ERC-165 probe and MUST NOT be used. If the probe reverts or runs out of gas, the predicate is treated as not advertising ERC-165: accept as best-effort.

Reviewer: (also has gas cap)

When a tool's `pricing` array specifies an onchain payment protocol, the payment flow is susceptible to replay attacks unless the protocol includes replay resistance. A malicious endpoint could replay a signed payment authorization to drain additional funds beyond what the user approved for a single invocation.

Payment protocols referenced in `pricing` MUST include replay protection. At minimum, each payment authorization MUST bind to a unique nonce or commitment that the payment contract enforces as single-use. Agents SHOULD use EIP-712 typed structured data for payment authorizations, including the `toolId`, a monotonic nonce, a `deadline` timestamp, and the chain ID in the signed payload. Agents MUST NOT sign open-ended approvals (e.g., unlimited ERC-20 `approve`) as a substitute for per-invocation payment authorizations.

Tool creators who define payment flows SHOULD document the replay-resistance mechanism in the manifest's `pricing[].protocol` description. Consumers that encounter a pricing entry without documented replay resistance SHOULD treat it as higher risk and SHOULD warn the user before proceeding.

### Malicious Endpoints

> Tool endpoints are creator-controlled URLs. The `endpoint` field MUST be an `https://` URL (see §2); consumers MUST reject manifests whose `endpoint` uses any other scheme. Any consumer that may cause a tool to be invoked on a user's behalf (agent frameworks, wallets, invocation proxies, or any surface exposing a "run this tool" affordance, using the same scoping as the 24-hour cache ceiling in [§7 Handling Verification Failure](#handling-verification-failure)) MUST reject endpoints that resolve to private IP ranges (RFC 1918, RFC 6598, loopback, link-local, IPv6 ULA `fc00::/7`, IPv6 link-local `fe80::/10`) to prevent SSRF attacks against internal services, and MUST resolve the host immediately before each invocation rather than caching the resolution, to prevent DNS rebinding. Purely informational surfaces that never invoke endpoints (e.g., indexer digests, historical registries) MAY use SHOULD-level rejection but SHOULD NOT relax it without an explicit policy. Agent frameworks that invoke tool endpoints MUST enforce request timeouts and response size limits.

Reviewer: SSRF not defined

The same defensive posture applies to the diagnostic `name()` views on both `IToolRegistry` and `IAccessPredicate`, and to `IToolRegistry.version()`: returns are deployer-controlled strings, so consumers MUST cap them at 256 bytes (matching the `label` cap in [§Access](#4-access)) and MUST treat over-limit returns as if the contract did not implement the function. This prevents a malicious registry or predicate from grieving discovery surfaces with a multi-megabyte string in what consumers expect to be a cheap diagnostic call.

### Manifest Parser Hardening

A creator controls the bytes served at `metadataURI`, and a permissionless registration makes every fetched manifest effectively attacker-controlled. A consumer that parses without limits can be DoSed by a single malicious registration. Consumers MUST enforce the following ceilings and MUST reject any manifest that exceeds them:

| Limit | Value | Why |
| --- | --- | --- |

> | Manifest byte size | 1 MiB (1,048,576 bytes) | Bounds HTTP body and parser memory; a fully-populated honest manifest with ten pricing entries and a moderately rich schema is well under 10 KiB. Consumers MUST truncate at the limit and treat oversize fetches as a verification failure. |

Reviewer: Nit: also define 10 KiB here

| `inputs` / `outputs` schema depth | 16 levels | Deep enough for any real JSON Schema composition (`anyOf` of tagged unions, nested objects); prevents stack exhaustion in recursive validators. |
| `inputs` + `outputs` total schema nodes | 1,024 nodes | Covers rich schemas; prevents pathological fan-out attacks that create millions of subschemas via shallow-but-wide structures. |

Each row above is an independent upper bound on its own axis; they do not jointly compose. The 1 MiB total-size cap is the binding constraint that limits the joint product. As a worked example, a manifest cannot simultaneously hit `access.requirements.length = 256` and per-entry `data = 4,096` decoded bytes: 256 maxed-out entries serialize to roughly 2.1 MiB of hex envelope (8,194 UTF-8 bytes per entry, plus label, kind, and JSON overhead), well over the 1 MiB ceiling. Consumers MUST enforce every row, but conformant manifests trade off depth for breadth and never hit every row at once.

> Regex `pattern` values inside embedded schemas MUST be evaluated by a matcher immune to catastrophic backtracking (e.g., RE2, or an implementation with a bounded step count). A matcher with exponential worst-case behavior on attacker input is unsafe and MUST NOT be used.

Reviewer: RE2?

### Rendering Manifest Content

Every string and URL in a fetched manifest is creator-controlled and MUST be treated as untrusted input by any surface that renders it. Consumers MUST apply defense-in-depth: validation at fetch time, contextual encoding at render time, and a restrictive Content Security Policy in the surrounding document.

The following normative rules apply to common render paths:

- **Text fields (`name`, `description`, `tags`).** Consumers MUST contextually encode these fields when rendering: HTML-escape for text nodes, attribute-escape for attributes, and JS-escape for script contexts. UIs that inject them into the DOM without encoding are vulnerable to stored XSS from a single malicious registration.
- **Markdown rendering of `description`.** Consumers that render Markdown MUST disable raw-HTML passthrough or sandbox the rendered output. A compliant Markdown renderer strips `<script>`, `<iframe>`, event handlers, and `javascript:` URLs, or runs under a CSP that neutralizes these surfaces.

> - **`image` URIs.** Consumers MUST NOT accept `javascript:`, `file:`, `data:text/html`, or `vbscript:` URIs for a tool icon: these schemes enable script execution or local-filesystem exposure and have no legitimate icon use case. Consumers SHOULD NOT accept `http:` (MITM risk) or `blob:` (cross-context leakage risk) URIs; consumers that do accept either MUST apply the scheme-appropriate defenses in the Per-Scheme Rendering Guidance below.

Reviewer: MITM abbreviation

- **`endpoint` and extension-namespaced URL fields as clickable links.** Consumers SHOULD NOT render `javascript:`, `file:`, or `blob:` schemes as links; `http:` links are permitted only with explicit user consent because of MITM risk; all external links SHOULD carry `rel="noopener noreferrer"` to prevent tab-nabbing.

#### Non-Normative: Per-Scheme Rendering Guidance for `image`

For any scheme consumers do accept, scheme-appropriate defenses should be layered on the normative rules above:

- `https://`: set `referrerpolicy="no-referrer"` and proxy through consumer-controlled infrastructure where feasible.

> - `ipfs://`: resolve through a trusted gateway the consumer operates, not an arbitrary public gateway. A hostile gateway can inject response headers, set tracking cookies, or serve manipulated content even when the CID is content-addressed.

Reviewer: CID abbreviation

---
eip: 8257
title: Agent Tool Registry
description: Minimal onchain registry for AI agent tools with extensible predicate-based access control
author: Cody Sears (@CodySearsOS), Ryan Ghods (@ryanio)
discussions-to: https://ethereum-magicians.org/t/erc-8257-agent-tool-registry/28457
status: Draft
type: Standards Track
category: ERC
created: 2026-04-17
requires: 165
---

## Abstract

> This ERC specifies a permissionless onchain registry for AI agent tools. Each registration commits a metadata URI and a content hash; invocation access is gated by an optional external predicate contract. Registrations are anchored to a canonical off-chain manifest through origin-binding (the manifest is served at a well-known path on the endpoint's origin) and creator self-attestation (the manifest declares which onchain address is entitled to register it). Pricing and access-model details are deferred: the manifest carries protocol-agnostic pricing hints, and access logic lives in the predicate layer.

Reviewer: ```suggestion
This ERC specifies a permissionless onchain registry for AI agent tools. Each registration commits a metadata URI and a content hash; invocation access is gated by an optional external predicate contract. Registrations are anchored to a canonical offchain manifest through origin-binding (the manifest is served at a well-known path on the endpoint's origin) and creator self-attestation (the manifest declares which onchain address is entitled to register it). Pricing and access-model details are deferred: the manifest carries protocol-agnostic pricing hints, and access logic lives in the predicate layer.
```

`offchain` is used in the entire document, so use it here too (another comment later where `off-chain` is used)

This ERC intentionally does not pin a maximum gas value for predicate calls. Gas costs of common predicate building blocks (SLOADs, hashing, signature recovery) change with hard forks (e.g., the SLOAD repricing anticipated under Glamsterdam, see [EIP-8038](./eip-8038.md)) and differ between L1 and the various L2 fee schedules; concrete gas figures also depend on the Solidity compiler version and optimizer settings used to build the predicate. Any fixed ceiling published in this document would either become too tight after a repricing (breaking predicates that worked before) or pre-emptively too loose. The gas-bound decision therefore belongs at the call site:

> - **Direct callers** (wallets and agent frameworks invoking `hasAccess` from off-chain or via `eth_call`) supply gas through the normal RPC budget. A predicate that exhausts the budget reverts; the caller treats this as "access denied" and MAY retry with a larger budget.

Reviewer: ```suggestion
- **Direct callers** (wallets and agent frameworks invoking `hasAccess` from offchain or via `eth_call`) supply gas through the normal RPC budget. A predicate that exhausts the budget reverts; the caller treats this as "access denied" and MAY retry with a larger budget.
```

When a tool's `pricing` array specifies an onchain payment protocol, the payment flow is susceptible to replay attacks unless the protocol includes replay resistance. A malicious endpoint could replay a signed payment authorization to drain additional funds beyond what the user approved for a single invocation.

Payment protocols referenced in `pricing` MUST include replay protection. At minimum, each payment authorization MUST bind to a unique nonce or commitment that the payment contract enforces as single-use. Agents SHOULD use EIP-712 typed structured data for payment authorizations, including the `toolId`, a monotonic nonce, a `deadline` timestamp, and the chain ID in the signed payload. Agents MUST NOT sign open-ended approvals (e.g., unlimited ERC-20 `approve`) as a substitute for per-invocation payment authorizations.

Tool creators who define payment flows SHOULD document the replay-resistance mechanism in the manifest's `pricing[].protocol` description. Consumers that encounter a pricing entry without documented replay resistance SHOULD treat it as higher risk and SHOULD warn the user before proceeding.

### Malicious Endpoints

> Tool endpoints are creator-controlled URLs. The `endpoint` field MUST be an `https://` URL (see §2); consumers MUST reject manifests whose `endpoint` uses any other scheme. Any consumer that may cause a tool to be invoked on a user's behalf (agent frameworks, wallets, invocation proxies, or any surface exposing a "run this tool" affordance, using the same scoping as the 24-hour cache ceiling in [§7 Handling Verification Failure](#handling-verification-failure)) MUST reject endpoints that resolve to private IP ranges (RFC 1918, RFC 6598, loopback, link-local, IPv6 ULA `fc00::/7`, IPv6 link-local `fe80::/10`) to prevent Server-Side Request Forgery (SSRF) attacks against internal services, and MUST resolve the host immediately before each invocation rather than caching the resolution, to prevent DNS rebinding. Purely informational surfaces that never invoke endpoints (e.g., indexer digests, historical registries) MAY use SHOULD-level rejection but SHOULD NOT relax it without an explicit policy. Agent frameworks that invoke tool endpoints MUST enforce request timeouts and response size limits.

Reviewer: ```suggestion
Tool endpoints are creator-controlled URLs. The `endpoint` field MUST be an `https://` URL (see §2); consumers MUST reject manifests whose `endpoint` uses any other scheme. Any consumer that may cause a tool to be invoked on a user's behalf (agent frameworks, wallets, invocation proxies, or any surface exposing a "run this tool" affordance, using the same scoping as the 24-hour cache ceiling in [§7 Handling Verification Failure](#handling-verification-failure)) MUST reject endpoints that resolve to private IP ranges ([RFC 1918](https://www.rfc-editor.org/rfc/rfc1918), [RFC 6598](https://www.rfc-editor.org/rfc/rfc6598), loopback, link-local, IPv6 ULA `fc00::/7`, IPv6 link-local `fe80::/10`) to prevent Server-Side Request Forgery (SSRF) attacks against internal services, and MUST resolve the host immediately before each invocation rather than caching the resolution, to prevent DNS rebinding. Purely informational surfaces that never invoke endpoints (e.g., indexer digests, historical registries) MAY use SHOULD-level rejection but SHOULD NOT relax it without an explicit policy. Agent frameworks that invoke tool endpoints MUST enforce request timeouts and response size limits.
```

Reviewer: (add RFC urls)
